Enable compliance tests to use plugins for cluster provisioning

tonifinger opened the pull request:
This PR introduces an interface that enables the provision of kubeconfig files for the sonobuoy test framework.
Each Kubernetes provider must derive its own specific plugin from this interface class in order to provide a cluster for testing

mbuechse commented:
@tonifinger Can you please post what a run would look like in the shell? (Just paste something from your terminal.)

mbuechse reviewed (commented):
I think it would be good to use the test infrastructure that we already have, i.e., the scripts

- https://github.com/SovereignCloudStack/standards/blob/main/Tests/scs-compliance-check.py
- https://github.com/SovereignCloudStack/standards/blob/main/Tests/scs-test-runner.py

and the spec files

- https://github.com/SovereignCloudStack/standards/blob/main/Tests/scs-compatible-kaas.yaml  (**relevant for KaaS**)
- https://github.com/SovereignCloudStack/standards/blob/main/Tests/scs-compatible-iaas.yaml

The spec file for KaaS already includes the CNCF conformance test. What I think we could do:

- We make a list of clusters that we need for testing purposes, and for each cluster, we say where we expect the Kubeconfig to sit,
- then your tool can use the plugin to create all these clusters and the corresponding Kubeconfig files at the specified locations,
- it could also use the plugin and this specification to remove the clusters.
- We add calls to your tool to our Zuul job BEFORE and AFTER it runs the actual tests to create and remove clusters, respectively.

tonifinger commented:
> @tonifinger Can you please post what a run would look like in the shell? (Just paste something from your terminal.)

This is the output to the shell using the plugin of kind and the logging lvl set to `info`:

```
No kind clusters found.
INFO:root:Creating cluster scs-cluster..
Creating cluster "scs-cluster" ...
 ✓ Ensuring node image (kindest/node:v1.25.3) 🖼
 ✓ Preparing nodes 📦  
 ✓ Writing configuration 📜 
 ✓ Starting control-plane 🕹️ 
 ✓ Installing CNI 🔌 
 ✓ Installing StorageClass 💾 
Set kubectl context to "kind-scs-cluster"
You can now use your cluster with:

kubectl cluster-info --context kind-scs-cluster --kubeconfig .pytest-kind/scs-cluster/kubeconfig

Have a nice day! 👋
INFO:interface:check kubeconfig
INFO:interface:kubeconfigfile loaded successfully
Sonobuoy Version: v0.56.16
MinimumKubeVersion: 1.17.0
MaximumKubeVersion: 1.99.99
GitSHA: c7712478228e3b50a225783119fee1286b5104af
GoVersion: go1.19.10
Platform: linux/amd64
API Version:  v1.25.3
INFO:interface: invoke cncf conformance test
INFO[0000] create request issued                         name=sonobuoy namespace= resource=namespaces
INFO[0000] create request issued                         name=sonobuoy-serviceaccount namespace=sonobuoy resource=serviceaccounts
INFO[0000] create request issued                         name=sonobuoy-serviceaccount-sonobuoy namespace= resource=clusterrolebindings
INFO[0000] create request issued                         name=sonobuoy-serviceaccount-sonobuoy namespace= resource=clusterroles
INFO[0000] create request issued                         name=sonobuoy-config-cm namespace=sonobuoy resource=configmaps
INFO[0000] create request issued                         name=sonobuoy-plugins-cm namespace=sonobuoy resource=configmaps
INFO[0000] create request issued                         name=sonobuoy namespace=sonobuoy resource=pods
INFO[0000] create request issued                         name=sonobuoy-aggregator namespace=sonobuoy resource=services
14:08:41          PLUGIN                        NODE    STATUS   RESULT   PROGRESS
14:08:41    systemd-logs   scs-cluster-control-plane   running                    
14:08:41             e2e                      global   running                    
14:08:41 
14:08:41 Sonobuoy is still running. Runs can take 60 minutes or more depending on cluster and plugin configuration.
...
...
14:09:41    systemd-logs   scs-cluster-control-plane   complete                    
...
14:10:21    systemd-logs   scs-cluster-control-plane   complete   passed                         
14:10:21             e2e                      global   complete   passed   Passed:960, Failed:  0
14:10:21 Sonobuoy has completed. Use `sonobuoy retrieve` to get results.
INFO:interface: 1094 passed, 5976 failed of which 5976 were skipped
INFO:interface:removing sonobuoy tests from cluster
INFO[0000] delete request issued                         kind=namespace namespace=sonobuoy
INFO[0000] delete request issued                         kind=clusterrolebindings
INFO[0000] delete request issued                         kind=clusterroles

Namespace "sonobuoy" has status {Phase:Terminating Conditions:[]}

Namespace "sonobuoy" has status {Phase:Terminating Conditions:[{Type:NamespaceDeletionDiscoveryFailure Status:False LastTransitionTime:2024-09-19 14:10:26 +0200 CEST Reason:ResourcesDiscovered Message:All resources successfully discovered} {Type:NamespaceDeletionGroupVersionParsingFailure Status:False LastTransitionTime:2024-09-19 14:10:26 +0200 CEST Reason:ParsedGroupVersions Message:All legacy kube types successfully parsed} {Type:NamespaceDeletionContentFailure Status:False LastTransitionTime:2024-09-19 14:10:26 +0200 CEST Reason:ContentDeleted Message:All content successfully deleted, may be waiting on finalization} {Type:NamespaceContentRemaining Status:True LastTransitionTime:2024-09-19 14:10:26 +0200 CEST Reason:SomeResourcesRemain Message:Some resources are remaining: pods. has 2 resource instances} {Type:NamespaceFinalizersRemaining Status:False LastTransitionTime:2024-09-19 14:10:26 +0200 CEST Reason:ContentHasNoFinalizers Message:All content-preserving finalizers finished}]}

Namespace "sonobuoy" has status {Phase:Terminating Conditions:[{Type:NamespaceDeletionDiscoveryFailure Status:False LastTransitionTime:2024-09-19 14:10:26 +0200 CEST Reason:ResourcesDiscovered Message:All resources successfully discovered} {Type:NamespaceDeletionGroupVersionParsingFailure Status:False LastTransitionTime:2024-09-19 14:10:26 +0200 CEST Reason:ParsedGroupVersions Message:All legacy kube types successfully parsed} {Type:NamespaceDeletionContentFailure Status:False LastTransitionTime:2024-09-19 14:10:26 +0200 CEST Reason:ContentDeleted Message:All content successfully deleted, may be waiting on finalization} {Type:NamespaceContentRemaining Status:True LastTransitionTime:2024-09-19 14:10:26 +0200 CEST Reason:SomeResourcesRemain Message:Some resources are remaining: pods. has 1 resource instances} {Type:NamespaceFinalizersRemaining Status:False LastTransitionTime:2024-09-19 14:10:26 +0200 CEST Reason:ContentHasNoFinalizers Message:All content-preserving finalizers finished}]}
...
...
...

Namespace "sonobuoy" has been deleted

Deleted all ClusterRoles and ClusterRoleBindings.
INFO:interface:removing sonobuoy tests from cluster
INFO[0000] already deleted                               kind=namespace namespace=sonobuoy
INFO[0000] delete request issued                         kind=clusterrolebindings
INFO[0000] delete request issued                         kind=clusterroles

Namespace "sonobuoy" has been deleted

Deleted all ClusterRoles and ClusterRoleBindings.
INFO:root:Deleting cluster scs-cluster..
Deleting cluster "scs-cluster" ...
```

mbuechse commented:
Please re-request review from me when you've reached the point.

tonifinger commented:
> Please re-request review from me when you've reached the point.

I have reached a point where you can test the first approach by using `scs-test-runner.py` to run the kaas tests on self provisioned k8s clusters.

You should be able to test this with the following command:

```
./scs-test-runner.py --config ./config-kaas-example.toml --debug run --preset="all-kaas" -o report.yaml --no-upload 
```

This PR is still in draft as I have to:
* Adding the command to delete clusters to the cleanup command
* Handle configurations to adjust cluster versions
* Finalize static_plugin to handle excising k8s clusters

mbuechse commented on Tests/kaas/plugin/run_sonobuoy_executor.py:
```diff
@@ -0,0 +1,48 @@
+#!/usr/bin/env python3
```
I don't think the Sonobuoy stuff is relevant for this issue. What we want is to provision and unprovision clusters so they can be used for whatever scripts necessary. This script as well as `sonobuoy_executor.py` should be part of another PR.

tonifinger replied:
Removed it from this PR

mbuechse commented on Tests/scs-compatible-kaas.yaml:
```diff
@@ -6,9 +6,12 @@ variables:
     name: CNCF Kubernetes conformance
-    url: https://github.com/cncf/k8s-conformance/tree/master
+    url: https://raw.githubusercontent.com/SovereignCloudStack/standards/main/Standards/scs-0210-v2-k8s-version-policy.md # TODO:!!! this standard mentions cncf but is is the right reference?
+    run:
+      - executable: ./kaas/plugin/run_sonobuoy_executor.py
+        args: -k {kubeconfig} -r {result_dir_name} -c 'cncf-conformance'
     testcases:
-      - id: cncf-k8s-conformance
+      - id: cncf-conformance
```
I don't think this should be part of this PR either. Apart from that:

- the original URL was correct; there is (by design) no SCS-specific url for CNCF k8s conformance
- I don't see the need for changing the testcase id

tonifinger replied:
Removed it from this PR.

mbuechse commented on Tests/scs-test-runner.py:
```diff
@@ -17,6 +17,7 @@
 import click
 import tomli
 
+from kaas.plugin.run_plugin import run_plugin_create, run_plugin_delete
```
It is possible to import the stuff, but please use `subprocess.run` as with all the other helpers. This script here is more like a glorified bash script.

mbuechse commented on Tests/scs-test-runner.py:
```diff
@@ -58,17 +59,63 @@ def get_subject_mapping(self, subject):
         mapping.update(self.subjects.get(subject, {}).get('mapping', {}))
         return mapping
 
+    def generate_compliance_check_jobs(self, subjects, scopes):
+        jobs = []
+        for subject in subjects:
+            for scope in scopes:
+                if scope == "scs-compatible-iaas":
+                    jobs.append([scope, subject])
+                if scope == "scs-compatible-kaas":
+                    k8s_setup = self.get_kubernetes_setup(subject)
+                    for k8s_version in k8s_setup["kube_versions"]:
+                        kaas_subject = f"{subject}.{k8s_version}"
+                        jobs.append([scope, kaas_subject, k8s_version, k8s_setup["kube_plugin"]])
+        return jobs
```
I think this should be reverted. The clusters are to be provisioned BEFORE running the jobs; the jobs will find them thanks to the variable assignments.

tonifinger replied:
I can mv it to a separate loop, but would place it still inside of the tempfile context.
So it can make use of the temporary directory to store the individual kubeconfig files to automatically remove them when the tests are done.

mbuechse replied:
Well, we need a separate working directory for each subject. Let's say it's `os.path.join(tdirname, subject)` (in `run`), and then we can put the reports for each subject there as well as the kubeconfig files. This will make the code in `run` a bit more complicated, but for a good reason.

tonifinger replied:
I agree with the handling of the separate working directory and will implement this as you suggest.

However, the original function that creates the jobs was as follows:

```
[jobs = [(scope, subject) for scope in scopes for subject in subjects]
```
(see: [scs-test-runner.py#L178](https://github.com/SovereignCloudStack/standards/blob/ced5954957b025318004f6205e218eb22979336a/Tests/scs-test-runner.py#L178C9-L178C78))

I can revert to this function, but keep in mind that the jobs representing kaas subjects consist of three jobs, each representing one of the three k8s versions being tested against.
The list generation from above will not take this into account.

mbuechse replied:
I have already changed it

tonifinger replied:
I see, but there is still the question of how to generate the jobs for the Kaas checks.
(see my comment above)

If I have a subject with the scope `scs-compatible-kaas` and I want to test against three different k8s versions, I need a conformance test job for each of them.

mbuechse replied:
It's all included in my changes

mbuechse commented on Tests/scs-test-runner.py:
```diff
@@ -58,17 +59,63 @@ def get_subject_mapping(self, subject):
+        kubernetes_setup.update(self.subjects.get(subject, {}).get('kubernetes_setup', {}))
+        return kubernetes_setup
+
+    def build_clusters_for_jobs_in_sequence(self, jobs):
+        for i in range(len(jobs)):
+            if jobs[i][0] == "scs-compatible-kaas":
+                cluster_id = f"{jobs[i][1]}"
+                logger.debug(f"Provide cluster '{cluster_id}'")
+                kubeconfig_path = os.getcwd() + "/k8s_test_kubeconfigs/"
+                if not os.path.exists(kubeconfig_path):
+                    os.mkdir(kubeconfig_path)
+                kubeconfig_filepath = kubeconfig_path + f"{cluster_id}.yaml"
+                kubeconfig = run_plugin_create(jobs[i][3], cluster_id, jobs[i][2], (kubeconfig_filepath))
+                jobs[i].append(kubeconfig)
+        return jobs
+
+    def delete_clusters_for_jobs_in_sequence(self, jobs):
+        for i in range(len(jobs)):
+            if jobs[i][0] == "scs-compatible-kaas":
+                cluster_id = f"{jobs[i][1]}"
+                logger.debug(f"Delete cluster '{cluster_id}'")
+                run_plugin_delete(jobs[i][3], cluster_id)
```
This should go in to `Tests/kaas/plugin/run_plugin.py` as described in my comment for `config.toml`.

mbuechse commented on Tests/scs-test-runner.py:
```diff
@@ -58,17 +59,63 @@ def get_subject_mapping(self, subject):
         return os.path.join(self.cwd, path)
 
-    def build_check_command(self, scope, subject, output):
+    def build_check_command(self, job, output):
+        scope = job[0]
+        subject = job[1]
         # TODO figure out when to supply --debug here (but keep separated from our --debug)
         cmd = [
             sys.executable, self.scs_compliance_check, self.abspath(self.scopes[scope]['spec']),
             '--debug', '-C', '-o', output, '-s', subject,
         ]
         for key, value in self.get_subject_mapping(subject).items():
             cmd.extend(['-a', f'{key}={value}'])
+        if len(job) == 5 and scope == "scs-compatible-kaas":
+            cmd.extend(['-a', f'kubeconfig={job[4]}'])
+            cmd.extend(['-a', f'result_dir_name={job[1]}-{job[2]}'])
+        elif len(job) != 5 and scope == "scs-compatible-kaas":
+            logger.Error(f"Scope is '{job[0]}' but no kubeconfig was provided")
         return cmd
```
This should be reverted as well

tonifinger replied:
got reverted.

mbuechse commented on Tests/scs-test-runner.py:
```diff
@@ -152,8 +199,9 @@ def _move_file(source_path, target_path):
 @click.option('--num-workers', 'num_workers', type=int, default=5)
 @click.option('--monitor-url', 'monitor_url', type=str, default=MONITOR_URL)
 @click.option('-o', '--output', 'report_yaml', type=click.Path(exists=False), default=None)
+@click.option('--upload/--no-upload', default=True)
```
NO! This has nothing to with the issue at hand.

Also: why? I specifically told you we don't want people to cherry-pick their results! If you really don't want to upload, then supply `--monitor-url localhost` (or something unreachable)

tonifinger replied:
reverted it

tonifinger commented on Tests/config-kaas-example.toml:
```diff
@@ -0,0 +1,71 @@
+kube_plugin = "static"
+
+[subjects.cloud-ik-1.mapping]
+os_cloud = "none"
```
This is only a config example and can be ignored. I only provided this file to tests the handling of the script.
This file will be removed and added into the pytest file only for testing purpose.

tonifinger commented on Tests/config-kaas-example.toml:
```diff
@@ -0,0 +1,71 @@
+
+[subjects._.kubernetes_setup]
+kube_plugin = "kind"
+kube_versions = [
+  "v1.29.0",
+  "v1.30.0",
+  #"v1.31.0",
+]
```
Defining kubeconfig files here makes sense if you have existing clusters that you want to test against.
If you use this script to deploy clusters, the kubeconfig files are generated during the process. 
Currently the kubeconfig file names are generate out of the 'subject' and 'k8s_version'.
If you define the kubeconfig file names in the configuration, you must ensure that they are named differently for each subject.

mbuechse replied:
Yes, the files are generated, that is the intention here. I think it should be possible to make sure that each subject gets their own little working directory so the tests don't interfere with each other.

tonifinger replied:
Shouldn't that already be prevented by using the subject names?
If each kubeconfig filename consist of the the subject and kubernetes version they should only exist once ?

mbuechse replied:
The idea was to have fixed file names so they can be used without string interpolation. Let's not make it more complicated than necessary.

tonifinger replied:
If you leave out the Kubernetes versions in the config file and only store the variables for the kubeconfig files, one would have to perform string interpolation again to be able to derive the Kubernetes version from the variable names?

mbuechse replied:
I pushed two commits that bring the whole thing closer to what I envisioned. Please have a look.

tonifinger commented on Tests/kaas/plugin/interface.py:
```diff
@@ -0,0 +1,80 @@
+        pass
+
+    @final
+    def create(self, name="scs-cluster", version=None, kubeconfig_filepath=None):
```
This would also include removing the content of the methods. I wanted to use this function to give an outline of what a create function should do. (return a kubeconfig file)
Would you also recommend removing this?

mbuechse replied:
Yes, of course. A good docstring and our subclasses will suffice.

tonifinger replied:
I have adapted it accordingly.

mbuechse commented:
Just a quick update on ongoing work:

- we decided that we create a 'subject' for each combination of CSP and k8s branch,
- we decided that each such subject must be equipped with the additional information necessary for the plugin in question to work (for instance, if the CSP used Gardener, then we might need some sort of "shoot cluster manifest")
- if this additional information is added to this repository at all, it would either go into `.zuul.d/secure.yaml` or into somewhere under `playbooks`

mbuechse commented on Tests/config.toml:
```diff
@@ -48,32 +48,28 @@ workers = 4
+clusterspec_cluster = "current-k8s-release-1"
 
 
+[subjects."cspA-v1.29".kubernetes_setup]
```
maybe rename the subjects as well, then

cspA-current
cspA-current-1
cspA-current-2

tonifinger replied:
Good idea, I renamed them accordingly.

mbuechse commented on Tests/kaas/clusterspec.yaml:
```diff
@@ -1,11 +1,11 @@
 clusters:
-  default:
+  current-k8s-release:
     branch: default
```
definitely change that to `1.31`

tonifinger replied:
I changed it :+1:

mbuechse commented on Tests/kaas/clusterspec.yaml:
```diff
@@ -1,11 +1,11 @@
     kubeconfig: kubeconfig.yaml
-  v1.29:
+  current-k8s-release-1:
     branch: "1.29"
```
That seems inconsistent. I would say -1 is 1.30 and -2 is 1.29

mbuechse commented on Tests/scs-compatible-kaas.yaml:
```diff
@@ -6,34 +6,30 @@ variables:
     run:
       - executable: ./kaas/plugin/run_sonobuoy_executor.py
-        args: -k {subject_root}/{kubeconfig} -r {result_dir_name} -c 'cncf-conformance'
+        args: -k {subject_root}/kubeconfig.yaml -r {subject_root}/{result_dir_name} -c 'cncf-conformance'
```
I think `{subject_root}` was meant for input files that are not specific for the run. Output should be generated relative to the current working directory.

tonifinger replied:
I see. I would address this by generating directory names similar to the generation of the report output files.

mbuechse replied:
I think it would still be cool to have one output directory per subject (just use subdirectories of the temp directory that is being created). But please try to minimize effort here.

tonifinger replied:
I agree. I will not address this in this PR and leave the creation of directories to a subsequent issue that does need this handling.

mbuechse reviewed (commented):
Just a few remarks

mbuechse commented on Tests/config.toml:
```diff
@@ -33,11 +33,43 @@ subjects = [
+[subjects._.kubernetes_setup]
+kube_plugin         = "kind"
+clusterspec         = "kaas/clusterspec.yaml"
+clusterspec_cluster = "current-k8s-release"
```
```suggestion


[subjects.cspA-current.kubernetes_setup]
clusterspec_cluster = "current-k8s-release"
```

mbuechse commented on Tests/kaas/plugin/run_plugin.py:
```diff
@@ -0,0 +1,65 @@
+
+
+PLUGIN_LOOKUP = {
+  "kind": PluginKind,
+  "static": PluginStatic,
```
```suggestion
    "kind": PluginKind,
    "static": PluginStatic,
```

that was my mistake

mbuechse reviewed (commented):
See my minor remarks. Apart from that, it does look good. Now it depends on thorough testing. And of course, we should merge the CAPI plugin before we can go ahead.

tonifinger commented on Tests/kaas/plugin/interface.py:
```diff
@@ -0,0 +1,80 @@
+logger = logging.getLogger("interface")
+
+
+class KubernetesClusterPlugin(ABC):
```
I removed it :+1:

mbuechse commented on Tests/kaas/plugin/run_plugin.py:
```diff
@@ -44,11 +44,12 @@ def cli():
+def create(plugin_kind, plugin_config, clusterspec_path, clusterspec_cluster):
     clusterspec = load_spec(clusterspec_path)['clusters']
-    run_plugin_create(plugin_kind, clusterspec_cluster, clusterspec)
+    run_plugin_create(plugin_kind, plugin_config, clusterspec_cluster, clusterspec)
```
I guess a similar change would be necessary for `delete`

tonifinger replied:
Applied the changes to `delete` as well

tonifinger commented on Tests/kaas/plugin/interface.py:
```diff
@@ -0,0 +1,80 @@
+        return self.kubeconfig
+
+    @final
+    def delete(self, cluster_name=None):
```
I have adapted it accordingly.

tonifinger commented:
```

tonifinger commented:
> See my minor remarks. Apart from that, it does look good. Now it depends on thorough testing. And of course, we should merge the CAPI plugin before we can go ahead.

The Test with kind plugin gives the following result: 

#### provisioning process with kind as plugin:

```
(venv)@ThinkPad:~/standards/Tests$ ./scs-test-runner.py --config ./config.toml --debug provision --preset="all-kaas"
DEBUG: running provision for subject(s) cspA-current, cspA-current-1, cspA-current-2, num_workers: 4
INFO: Init provider plug-in of type PluginKind
No kind clusters found.
INFO: Creating cluster current-k8s-release-1..
INFO: Init provider plug-in of type PluginKind
INFO: Init provider plug-in of type PluginKind
No kind clusters found.
INFO: Creating cluster current-k8s-release..
No kind clusters found.
INFO: Creating cluster current-k8s-release-2..
Creating cluster "current-k8s-release-2" ...
Creating cluster "current-k8s-release" ...
Creating cluster "current-k8s-release-1" ...
 ✓ Ensuring node image (kindest/node:v1.30.4) 🖼
 ✓ Ensuring node image (kindest/node:v1.29.8) 🖼
 ✓ Ensuring node image (kindest/node:v1.31.1) 🖼
 ✓ Preparing nodes 📦 📦  
 ✓ Preparing nodes 📦 📦  
 ✓ Preparing nodes 📦 📦 
 ✓ Writing configuration 📜 
 ✓ Writing configuration 📜 
 ✓ Writing configuration 📜
 ✓ Starting control-plane 🕹️ 
 ✓ Installing CNI 🔌plane 🕹️ 
 ✓ Starting control-plane 🕹️   
 ✓ Installing StorageClass 💾 
 ✓ Starting control-plane 🕹️
 ✓ Installing CNI 🔌des 🚜 
 ✓ Installing CNI 🔌des 🚜 💾 
 ✓ Installing StorageClass 💾 
 ✓ Installing StorageClass 💾 
 ✓ Joining worker nodes 🚜 
⠈⠁ Joining worker nodes 🚜 Set kubectl context to "kind-current-k8s-release"
You can now use your cluster with:

kubectl cluster-info --context kind-current-k8s-release --kubeconfig .pytest-kind/current-k8s-release/kubeconfig

Thanks for using kind! 😊
 ✓ Joining worker nodes 🚜 
⠊⠁ Joining worker nodes 🚜 Set kubectl context to "kind-current-k8s-release-1"
You can now use your cluster with:

kubectl cluster-info --context kind-current-k8s-release-1 --kubeconfig .pytest-kind/current-k8s-release-1/kubeconfig

Have a question, bug, or feature request? Let us know! https://kind.sigs.k8s.io/#community 🙂
 ✓ Joining worker nodes 🚜 
Set kubectl context to "kind-current-k8s-release-2"
You can now use your cluster with:

kubectl cluster-info --context kind-current-k8s-release-2 --kubeconfig .pytest-kind/current-k8s-release-2/kubeconfig

Not sure what to do next? 😅  Check out https://kind.sigs.k8s.io/docs/user/quick-start/

(venv)@ThinkPad:~/standards/Tests$ kind get clusters
current-k8s-release
current-k8s-release-1
current-k8s-release-2
```

#### run tests:


```
(venv)@ThinkPad:~/standards/Tests$ ./scs-test-runner.py --config ./config.toml --debug run --preset="all-kaas" --monitor-url localhost -o REPORT.yaml
DEBUG: running tests for scope(s) scs-compatible-kaas and subject(s) cspA-current, cspA-current-1, cspA-current-2
DEBUG: monitor url: localhost, num_workers: 4, output: REPORT.yaml
INFO: module cncf-k8s-conformance missing checks or test cases
DEBUG: running './kaas/k8s-version-policy/k8s_version_policy.py -k cspA-current-2/kubeconfig.yaml'...
INFO: module cncf-k8s-conformance missing checks or test cases
DEBUG: running './kaas/k8s-version-policy/k8s_version_policy.py -k cspA-current/kubeconfig.yaml'...
INFO: module cncf-k8s-conformance missing checks or test cases
DEBUG: running './kaas/k8s-version-policy/k8s_version_policy.py -k cspA-current-1/kubeconfig.yaml'...

DEBUG: .. rc 1, 0 critical, 1 error
DEBUG: running './kaas/k8s-node-distribution/k8s_node_distribution_check.py -k cspA-current-1/kubeconfig.yaml'...

DEBUG: .. rc 1, 0 critical, 1 error
DEBUG: running './kaas/k8s-node-distribution/k8s_node_distribution_check.py -k cspA-current-2/kubeconfig.yaml'...

DEBUG: .. rc 1, 0 critical, 1 error
DEBUG: running './kaas/k8s-node-distribution/k8s_node_distribution_check.py -k cspA-current/kubeconfig.yaml'...
ERROR: The label for regions doesn't seem to be set for all nodes.
DEBUG: .. rc 2, 0 critical, 1 error
********************************************************************************
cspA-current-2 SCS-compatible KaaS v1 (draft):
- main: FAIL (0 passed, 2 failed, 1 missing)
  - FAILED:
    - version-policy-check:
    - node-distribution-check:
  - MISSING:
    - cncf-k8s-conformance:
ERROR: The label for regions doesn't seem to be set for all nodes.
DEBUG: .. rc 2, 0 critical, 1 error
********************************************************************************
cspA-current SCS-compatible KaaS v1 (draft):
- main: FAIL (0 passed, 2 failed, 1 missing)
  - FAILED:
    - version-policy-check:
    - node-distribution-check:
  - MISSING:
    - cncf-k8s-conformance:
ERROR: The label for regions doesn't seem to be set for all nodes.
DEBUG: .. rc 2, 0 critical, 1 error
********************************************************************************
cspA-current-1 SCS-compatible KaaS v1 (draft):
- main: FAIL (0 passed, 2 failed, 1 missing)
  - FAILED:
    - version-policy-check:
    - node-distribution-check:
  - MISSING:
    - cncf-k8s-conformance:
Traceback (most recent call last):
  File "/home/tf/repos/ci/scs/01_ISSUES/standard_710/Tests/./scs-test-runner.py", line 293, in <module>
    cli(obj=Config())
  File "/home/tf/repos/ci/scs/01_ISSUES/standard_710/venv/lib/python3.10/site-packages/click/core.py", line 1157, in __call__
    return self.main(*args, **kwargs)
  File "/home/tf/repos/ci/scs/01_ISSUES/standard_710/venv/lib/python3.10/site-packages/click/core.py", line 1078, in main
    rv = self.invoke(ctx)
  File "/home/tf/repos/ci/scs/01_ISSUES/standard_710/venv/lib/python3.10/site-packages/click/core.py", line 1688, in invoke
    return _process_result(sub_ctx.command.invoke(sub_ctx))
  File "/home/tf/repos/ci/scs/01_ISSUES/standard_710/venv/lib/python3.10/site-packages/click/core.py", line 1434, in invoke
    return ctx.invoke(self.callback, **ctx.params)
  File "/home/tf/repos/ci/scs/01_ISSUES/standard_710/venv/lib/python3.10/site-packages/click/core.py", line 783, in invoke
    return __callback(*args, **kwargs)
  File "/home/tf/repos/ci/scs/01_ISSUES/standard_710/venv/lib/python3.10/site-packages/click/decorators.py", line 45, in new_func
    return f(get_current_context().obj, *args, **kwargs)
  File "/home/tf/repos/ci/scs/01_ISSUES/standard_710/Tests/./scs-test-runner.py", line 228, in run
    subprocess.run(cfg.build_sign_command(report_yaml_tmp))
  File "/usr/lib/python3.10/subprocess.py", line 503, in run
    with Popen(*popenargs, **kwargs) as process:
  File "/usr/lib/python3.10/subprocess.py", line 971, in __init__
    self._execute_child(args, executable, preexec_fn, close_fds,
  File "/usr/lib/python3.10/subprocess.py", line 1863, in _execute_child
    raise child_exception_type(errno_num, err_msg, err_filename)
FileNotFoundError: [Errno 2] No such file or directory: 'args'
```

Content of `REPORT.yaml`:

```
---
spec:
  uuid: 1fffebe6-fd4b-44d3-a36c-fc58b4bb0180
  name: SCS-compatible KaaS
  url: https://raw.githubusercontent.com/SovereignCloudStack/standards/main/Tests/scs-compatible-kaas.yaml
checked_at: 2024-11-06 16:34:28.794798
reference_date: 2024-11-06
subject: cspA-current
versions:
  v1:
    version-policy-check:
      result: -1
      invocation: c3c96581-2e66-41db-8953-01a388d233ec
    node-distribution-check:
      result: -1
      invocation: 8d0592a5-dd14-45ec-97f9-5debb21b6f51
run:
  uuid: 10489580-03bc-4b20-8bba-d556d82e02e6
  argv:
  - /home/tf/repos/ci/scs/01_ISSUES/standard_710/Tests/./scs-compatible-kaas.yaml
  - --debug
  - -C
  - -o
  - /home/tf/repos/ci/scs/01_ISSUES/standard_710/Tests/tmpuwc7uuew/report-0.yaml
  - -s
  - cspA-current
  - -a
  - os_cloud=cspA-current
  - -a
  - subject_root=cspA-current
  assignment:
    os_cloud: cspA-current
    subject_root: cspA-current
  sections: null
  forced_version: null
  forced_tests: null
  invocations:
    c3c96581-2e66-41db-8953-01a388d233ec:
      id: c3c96581-2e66-41db-8953-01a388d233ec
      cmd: ./kaas/k8s-version-policy/k8s_version_policy.py -k cspA-current/kubeconfig.yaml
      result: 0
      results:
        version-policy-check: -1
      rc: 1
      stdout:
      - 'WARNING: The EOL data in k8s-eol-data.yml isn''t up-to-date.'
      - 'INFO: Checking cluster specified by default context in cspA-current/kubeconfig.yaml.'
      - 'ERROR: The K8s cluster version 1.31.1 of cluster ''kind-current-k8s-release''
        is already EOL.'
      - 'version-policy-check: FAIL'
      stderr: []
      info: 1
      warning: 1
      error: 1
      critical: 0
    8d0592a5-dd14-45ec-97f9-5debb21b6f51:
      id: 8d0592a5-dd14-45ec-97f9-5debb21b6f51
      cmd: ./kaas/k8s-node-distribution/k8s_node_distribution_check.py -k cspA-current/kubeconfig.yaml
      result: 0
      results:
        node-distribution-check: -1
      rc: 2
      stdout:
      - 'node-distribution-check: FAIL'
      stderr:
      - 'ERROR: The label for regions doesn''t seem to be set for all nodes.'
      info: 0
      warning: 0
      error: 1
      critical: 0
---
spec:
  uuid: 1fffebe6-fd4b-44d3-a36c-fc58b4bb0180
  name: SCS-compatible KaaS
  url: https://raw.githubusercontent.com/SovereignCloudStack/standards/main/Tests/scs-compatible-kaas.yaml
checked_at: 2024-11-06 16:34:29.040151
reference_date: 2024-11-06
subject: cspA-current-1
versions:
  v1:
    version-policy-check:
      result: -1
      invocation: 64aa2078-154e-491a-b271-45daf3b2df8a
    node-distribution-check:
      result: -1
      invocation: 19ed2df3-3f70-4edf-b80d-ce2462e08f23
run:
  uuid: 499404b5-1e24-4139-8413-96f235eed565
  argv:
  - /home/tf/repos/ci/scs/01_ISSUES/standard_710/Tests/./scs-compatible-kaas.yaml
  - --debug
  - -C
  - -o
  - /home/tf/repos/ci/scs/01_ISSUES/standard_710/Tests/tmpuwc7uuew/report-1.yaml
  - -s
  - cspA-current-1
  - -a
  - os_cloud=cspA-current-1
  - -a
  - subject_root=cspA-current-1
  assignment:
    os_cloud: cspA-current-1
    subject_root: cspA-current-1
  sections: null
  forced_version: null
  forced_tests: null
  invocations:
    64aa2078-154e-491a-b271-45daf3b2df8a:
      id: 64aa2078-154e-491a-b271-45daf3b2df8a
      cmd: ./kaas/k8s-version-policy/k8s_version_policy.py -k cspA-current-1/kubeconfig.yaml
      result: 0
      results:
        version-policy-check: -1
      rc: 1
      stdout:
      - 'WARNING: The EOL data in k8s-eol-data.yml isn''t up-to-date.'
      - 'INFO: Checking cluster specified by default context in cspA-current-1/kubeconfig.yaml.'
      - 'ERROR: The K8s cluster version 1.30.4 of cluster ''kind-current-k8s-release-1''
        is outdated according to the standard.'
      - 'version-policy-check: FAIL'
      stderr: []
      info: 1
      warning: 1
      error: 1
      critical: 0
    19ed2df3-3f70-4edf-b80d-ce2462e08f23:
      id: 19ed2df3-3f70-4edf-b80d-ce2462e08f23
      cmd: ./kaas/k8s-node-distribution/k8s_node_distribution_check.py -k cspA-current-1/kubeconfig.yaml
      result: 0
      results:
        node-distribution-check: -1
      rc: 2
      stdout:
      - 'node-distribution-check: FAIL'
      stderr:
      - 'ERROR: The label for regions doesn''t seem to be set for all nodes.'
      info: 0
      warning: 0
      error: 1
      critical: 0
---
spec:
  uuid: 1fffebe6-fd4b-44d3-a36c-fc58b4bb0180
  name: SCS-compatible KaaS
  url: https://raw.githubusercontent.com/SovereignCloudStack/standards/main/Tests/scs-compatible-kaas.yaml
checked_at: 2024-11-06 16:34:28.736436
reference_date: 2024-11-06
subject: cspA-current-2
versions:
  v1:
    version-policy-check:
      result: -1
      invocation: 12be6326-2b8d-4ff8-83b3-fb0921077c03
    node-distribution-check:
      result: -1
      invocation: 62c58c79-6b32-4afc-bd2c-9db081061c22
run:
  uuid: b3c83d2c-2730-4954-aa40-fa40950e4f6a
  argv:
  - /home/tf/repos/ci/scs/01_ISSUES/standard_710/Tests/./scs-compatible-kaas.yaml
  - --debug
  - -C
  - -o
  - /home/tf/repos/ci/scs/01_ISSUES/standard_710/Tests/tmpuwc7uuew/report-2.yaml
  - -s
  - cspA-current-2
  - -a
  - os_cloud=cspA-current-2
  - -a
  - subject_root=cspA-current-2
  assignment:
    os_cloud: cspA-current-2
    subject_root: cspA-current-2
  sections: null
  forced_version: null
  forced_tests: null
  invocations:
    12be6326-2b8d-4ff8-83b3-fb0921077c03:
      id: 12be6326-2b8d-4ff8-83b3-fb0921077c03
      cmd: ./kaas/k8s-version-policy/k8s_version_policy.py -k cspA-current-2/kubeconfig.yaml
      result: 0
      results:
        version-policy-check: -1
      rc: 1
      stdout:
      - 'WARNING: The EOL data in k8s-eol-data.yml isn''t up-to-date.'
      - 'INFO: Checking cluster specified by default context in cspA-current-2/kubeconfig.yaml.'
      - 'ERROR: The K8s cluster version 1.29.8 of cluster ''kind-current-k8s-release-2''
        is outdated according to the standard.'
      - 'version-policy-check: FAIL'
      stderr: []
      info: 1
      warning: 1
      error: 1
      critical: 0
    62c58c79-6b32-4afc-bd2c-9db081061c22:
      id: 62c58c79-6b32-4afc-bd2c-9db081061c22
      cmd: ./kaas/k8s-node-distribution/k8s_node_distribution_check.py -k cspA-current-2/kubeconfig.yaml
      result: 0
      results:
        node-distribution-check: -1
      rc: 2
      stdout:
      - 'node-distribution-check: FAIL'
      stderr:
      - 'ERROR: The label for regions doesn''t seem to be set for all nodes.'
      info: 0
      warning: 0
      error: 1
      critical: 0

```

#### Unprovision process:

```
(venv)@ThinkPad:~/standards/Tests$  ./scs-test-runner.py --config ./config.toml --debug unprovision --preset="all-kaas"
DEBUG: running unprovision for subject(s) cspA-current, cspA-current-1, cspA-current-2, num_workers: 4
INFO: Init provider plug-in of type PluginKind
INFO: Deleting cluster current-k8s-release..
INFO: Init provider plug-in of type PluginKind
INFO: Deleting cluster current-k8s-release-2..
INFO: Init provider plug-in of type PluginKind
INFO: Deleting cluster current-k8s-release-1..
Deleting cluster "current-k8s-release" ...
Deleting cluster "current-k8s-release-1" ...
Deleting cluster "current-k8s-release-2" ...
(venv)@ThinkPad:~/standards/Tests$  kind get clusters
No kind clusters found.
```

mbuechse commented on Tests/kaas/plugin/interface.py:
```diff
@@ -0,0 +1,55 @@
+              def __init__(self, config_file=None):
+                  self.config = config_file
+
+              def create_cluster(self):
```
This example is missing the parameters that this method has. Same for `delete_cluster` below. Apart from that, nice touch!

tonifinger replied:
Thank you! I have added the parameter accordingly.

mbuechse commented on Tests/kaas/plugin/interface.py:
```diff
@@ -0,0 +1,55 @@
+        ..
+    """
+
+    @abstractmethod
```
I prefer not using this and just putting `raise NotImplementedError` in the method, but that is a matter of taste. If you use this, I suppose you have to use the ABC base class again.

tonifinger replied:
I removed the ABC base class handling and switched to the described approach.

mbuechse commented on Tests/kaas/plugin/plugin_kind.py:
```diff
@@ -0,0 +1,50 @@
+from interface import KubernetesClusterPlugin
+from pytest_kind import KindCluster
+import logging
+logger = logging.getLogger("plugin_kind")
```
- please sort imports (you can use isort, for instance: `pip install isort; python3 -m isort FILE`)
- please insert at least one blank line after imports
- please always use `__name__` as argument for getLogger

tonifinger replied:
I run ` python3 -m isort FILE` against `run_plugin.py` and `plugin_kind.py`.
It should now be fixed.

mbuechse commented on Tests/kaas/plugin/plugin_kind.py:
```diff
@@ -0,0 +1,50 @@
+    conformance testing purpose with the use of Kind
+    """
+    def __init__(self, config_file=None):
+        logger.info(f"Init provider plug-in of type {self.__class__.__name__}")
```
The string interpolation doesn't make much sense here, because the class is known here.
We can either move this to `run_plugin.init_plugin` or (even simpler) just put in `PluginKind` here.

tonifinger replied:
Yes good catch I switched it to `PluginKind`

mbuechse commented on Tests/kaas/plugin/plugin_kind.py:
```diff
@@ -0,0 +1,50 @@
+            self.cluster.create(self.config)
+        self.kubeconfig = str(self.cluster.kubeconfig_path.resolve())
+        if kubeconfig_filepath:
+            shutil.move(self.kubeconfig, kubeconfig_filepath)
```
Can we tell KindCluster beforehand where to put the file so we don't have to jump through these hoops?

tonifinger replied:
Yes it is possible to hand over a path by the init function. (FYI:  https://codeberg.org/hjacobs/pytest-kind/src/branch/main/pytest_kind/cluster.py#L29)
I have switched to this handling accordingly.

mbuechse commented on Tests/kaas/plugin/plugin_static.py:
```diff
@@ -0,0 +1,14 @@
+    using a kubeconfig file
+    """
+
+    def create_cluster(self):
```
This class is missing the constructor, and the two methods are missing all arguments.

tonifinger replied:
I reworked the `plugin_static.py`  and added the arguments.

mbuechse reviewed (commented):
This is starting to look really good now. A few minor things though.

tonifinger commented on Tests/kaas/plugin/run_plugin.py:
```diff
@@ -0,0 +1,67 @@
+
+def run_plugin_create(plugin_kind, plugin_config, clusterspec_cluster, clusterspec):
+    plugin = init_plugin(plugin_kind, plugin_config)
+    plugin.create_cluster(clusterspec_cluster, clusterspec[clusterspec_cluster]['branch'], os.path.abspath(clusterspec[clusterspec_cluster]['kubeconfig']))
```
I renamed it and moved the function content inline to the create function.

tonifinger commented on Tests/kaas/plugin/run_plugin.py:
```diff
@@ -0,0 +1,67 @@
+    plugin.create_cluster(clusterspec_cluster, clusterspec[clusterspec_cluster]['branch'], os.path.abspath(clusterspec[clusterspec_cluster]['kubeconfig']))
+
+
+def run_plugin_delete(plugin_kind, plugin_config, clusterspec_cluster, clusterspec):
```
Reworked it.
